internal/route/api/v1/repo/contents.go

```diff
@@ -0,0 +1,172 @@
+package repo
+
+import (
+	"encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"net/url"
+
+	"github.com/gogs/git-module"
+
+	"gogs.io/gogs/internal/context"
+)
+
+type repoContent struct {
+	Type            string `json:"type"`
+	Target          string `json:"target,omitempty"`
+	SubmoduleGitURL string `json:"submodule_git_url,omitempty"`
+	Encoding        string `json:"encoding,omitempty"`
+	Size            int64  `json:"size"`
+	Name            string `json:"name"`
+	Path            string `json:"path"`
+	Content         string `json:"content,omitempty"`
+	Sha             string `json:"sha"`
+	URL             string `json:"url"`
+	GitURL          string `json:"git_url"`
+	HTMLURL         string `json:"html_url"`
+	DownloadURL     string `json:"download_url"`
+	Links           Links  `json:"_links"`
+}
+
+type Links struct {
+	Git  string `json:"git"`
+	Self string `json:"self"`
+	HTML string `json:"html"`
+}
+
+func GetContents(c *context.APIContext) {
+	treeEntry, err := c.Repo.Commit.GetTreeEntryByPath(c.Repo.TreePath)
+	if err != nil {
+		c.NotFoundOrServerError("GetTreeEntryByPath", git.IsErrNotExist, err)
+		return
+	}
+	username := c.Params(":username")
+	reponame := c.Params(":reponame")
+
+	// TODO: figure out the best way to do this
+	// :base-url/:username/:project/raw/:refs/:path
+	templateDownloadURL := "%s/%s/%s/raw/%s"
+	// :base-url/repos/:username/:project/contents/:path
+	templateSelfLink := "%s/repos/%s/%s/contents/%s"
+	// :baseurl/repos/:username/:project/git/trees/:sha
+	templateGitURLLink := "%s/repos/%s/%s/trees/%s"
+	// :baseurl/repos/:username/:project/tree/:sha
+	templateHTMLLLink := "%s/repos/%s/%s/tree/%s"
+
+	gitURL := fmt.Sprintf(templateGitURLLink, c.BaseURL, username, reponame, treeEntry.ID.String())
+	htmlURL := fmt.Sprintf(templateHTMLLLink, c.BaseURL, username, reponame, treeEntry.ID.String())
+	selfURL := fmt.Sprintf(templateSelfLink, c.BaseURL, username, reponame, c.Repo.TreePath)
+
+	contents := &repoContent{
+		Size:        treeEntry.Size(),
+		Name:        treeEntry.Name(),
+		Path:        c.Repo.TreePath,
+		Sha:         treeEntry.ID.String(),
+		URL:         selfURL,
+		GitURL:      gitURL,
+		HTMLURL:     htmlURL,
+		DownloadURL: fmt.Sprintf(templateDownloadURL, c.BaseURL, username, reponame, c.Repo.TreePath),
+		Links: Links{
+			Git:  gitURL,
+			Self: selfURL,
+			HTML: htmlURL,
+		},
+	}
+
+	// A tree entry can only be one of the following types:
+	// 1. Tree (Directory)
+	// 2. SubModule
+	// 3. SymLink
+	// 4. Blob
+	if treeEntry.IsSubModule() {
+		contents.Type = "submodule"
+		parsedURL, err := url.Parse(c.BaseURL)
+		if err != nil {
+			c.ServerError("ErrorURLParse", err)
+			return
+		}
+		host := parsedURL.Host
+		contents.SubmoduleGitURL = fmt.Sprintf("git://%s/%s/%s.git", host, username, reponame)
+		c.JSONSuccess(contents)
+		return
+
+	} else if treeEntry.IsLink() {
+		contents.Type = "symlink"
+		contents.Target = c.Repo.TreePath
```

Comment:
Same in dirTree.

Reply:
More specifically, according to the Github API v3: `"target": "/path/to/symlink/target"`. Target is the `blob.Data` if the target is a file - response will be similar to a blob data.

If the target is a directory, then looks like, we would need to just set Target to the path of the dir that is requested and the current format of response seems ok to me.

I will work on it, thanks!

Reply:
For an example in my test repo: https://github.com/bharatnc/test-repo dir2 is a symlink to dir1
```bash
curl -H "Authorization: token REDACTED" https://api.github.com/repos/bharatnc/test-repo/contents/dir1 | jq .

[
  {
    "name": "file1.txt",
    "path": "dir1/file1.txt",
    "sha": "e69de29bb2d1d6434b8b29ae775ad8c2e48c5391",
    "size": 0,
    "url": "https://api.github.com/repos/bharatnc/test-repo/contents/dir1/file1.txt?ref=master",
    "html_url": "https://github.com/bharatnc/test-repo/blob/master/dir1/file1.txt",
    "git_url": "https://api.github.com/repos/bharatnc/test-repo/git/blobs/e69de29bb2d1d6434b8b29ae775ad8c2e48c5391",
    "download_url": "https://raw.githubusercontent.com/bharatnc/test-repo/master/dir1/file1.txt",
    "type": "file",
    "_links": {
      "self": "https://api.github.com/repos/bharatnc/test-repo/contents/dir1/file1.txt?ref=master",
      "git": "https://api.github.com/repos/bharatnc/test-repo/git/blobs/e69de29bb2d1d6434b8b29ae775ad8c2e48c5391",
      "html": "https://github.com/bharatnc/test-repo/blob/master/dir1/file1.txt"
    }
  }
]

curl -H "Authorization: token REDACTED" https://api.github.com/repos/bharatnc/test-repo/contents/dir2 | jq .

{
  "name": "dir2",
  "path": "dir2",
  "sha": "df490f837a85cbc0e94a6921286c2f97b8dde399",
  "size": 4,
  "url": "https://api.github.com/repos/bharatnc/test-repo/contents/dir2?ref=master",
  "html_url": "https://github.com/bharatnc/test-repo/blob/master/dir2",
  "git_url": "https://api.github.com/repos/bharatnc/test-repo/git/blobs/df490f837a85cbc0e94a6921286c2f97b8dde399",
  "download_url": "https://raw.githubusercontent.com/bharatnc/test-repo/master/dir2",
  "type": "symlink",
  "target": "dir1",
  "_links": {
    "self": "https://api.github.com/repos/bharatnc/test-repo/contents/dir2?ref=master",
    "git": "https://api.github.com/repos/bharatnc/test-repo/git/blobs/df490f837a85cbc0e94a6921286c2f97b8dde399",
    "html": "https://github.com/bharatnc/test-repo/blob/master/dir2"
  }
}
```
file3 is a symlink to file 1, but the response is as follows:

```
curl -H "Authorization: token REDACTED" https://api.github.com/repos/bharatnc/test-repo/contents/file3.txt | jq .
{
  "name": "file3.txt",
  "path": "file3.txt",
  "sha": "e69de29bb2d1d6434b8b29ae775ad8c2e48c5391",
  "size": 0,
  "url": "https://api.github.com/repos/bharatnc/test-repo/contents/file3.txt?ref=master",
  "html_url": "https://github.com/bharatnc/test-repo/blob/master/file3.txt",
  "git_url": "https://api.github.com/repos/bharatnc/test-repo/git/blobs/e69de29bb2d1d6434b8b29ae775ad8c2e48c5391",
  "download_url": "https://raw.githubusercontent.com/bharatnc/test-repo/master/file3.txt",
  "type": "file",
  "content": "",
  "encoding": "base64",
  "_links": {
    "self": "https://api.github.com/repos/bharatnc/test-repo/contents/file3.txt?ref=master",
    "git": "https://api.github.com/repos/bharatnc/test-repo/git/blobs/e69de29bb2d1d6434b8b29ae775ad8c2e48c5391",
    "html": "https://github.com/bharatnc/test-repo/blob/master/file3.txt"
  }
}

curl -H "Authorization: token REDACTED" https://api.github.com/repos/bharatnc/test-repo/contents/file1.txt | jq .

{
  "name": "file1.txt",
  "path": "file1.txt",
  "sha": "e69de29bb2d1d6434b8b29ae775ad8c2e48c5391",
  "size": 0,
  "url": "https://api.github.com/repos/bharatnc/test-repo/contents/file1.txt?ref=master",
  "html_url": "https://github.com/bharatnc/test-repo/blob/master/file1.txt",
  "git_url": "https://api.github.com/repos/bharatnc/test-repo/git/blobs/e69de29bb2d1d6434b8b29ae775ad8c2e48c5391",
  "download_url": "https://raw.githubusercontent.com/bharatnc/test-repo/master/file1.txt",
  "type": "file",
  "content": "",
  "encoding": "base64",
  "_links": {
    "self": "https://api.github.com/repos/bharatnc/test-repo/contents/file1.txt?ref=master",
    "git": "https://api.github.com/repos/bharatnc/test-repo/git/blobs/e69de29bb2d1d6434b8b29ae775ad8c2e48c5391",
    "html": "https://github.com/bharatnc/test-repo/blob/master/file1.txt"
  }
}
```

*Note:* The file content is missing as they are all empty.

Reply:
@unknwon  so based on the above behavior of the Gitlab api, it makes me think,
1.  A `blob` even if it's a `symlink` is still treated like a `blob`. Meaning, you still display the file contents with the same response that's for a `blob`.
2. Only a `dir`, if it's of type `symlink` responds with json with `target` set to the actual file name and doesn't display actually display the file tree in the response.

So ideally, we only need this check if the entry is a `dir`:

```golang
else if treeEntry.IsLink() {
		contents.Type = "symlink"
		contents.Target = c.Repo.TreePath
		c.JSONSuccess(contents)
		return

	}
```
If the above check reveals that a `dir` is also a `symlink`, then instead of displaying the entires, we just respond with the json pointing to the **"path of parent"** as `target`.

One **question** that I wanted to ask you is if git-module currently has methods to get the parent path/name for a given symlink?

As far as I have looked, I didn't find any.  If it's there, we can conveniently set that and respond with a valid json. Thanks and also let me know what you think?

Reply:
> If the above check reveals that a `dir` is also a `symlink`, then instead of displaying the entires, we just respond with the json pointing to the **"path of parent"** as `target`.

What do you mean by "path of parent"? 

---

The example you gave for "dir2 is a symlink to dir1" is actually request "dir1" not "dir2".

Reply:
> > If the above check reveals that a `dir` is also a `symlink`, then instead of displaying the entires, we just respond with the json pointing to the **"path of parent"** as `target`.
> 
> What do you mean by "path of parent"?
> 
> The example you gave for "dir2 is a symlink to dir1" is actually requesting "dir1" not "dir2".

True, sorry, I have corrected myself above, my question was -  if dir2 is symlink to dir1, and we request for dir2,  is it currently possible to know that dir1 is infact the original file  (that points to dir2) ?

Reply:
Symlink is basically a TXT file with special bits, so the content of a symlink file is where it points to.

Reply:
Oh I see! Thank makes perfect sense :). Pardon me for the many redirections, I think that's exactly what you meant in your very first comment in this thread!

> It is not true, it is a symlink so the target is actually the file content (i.e. blob.Data())

Will change the Target to be the blob.Data()

Reply:
Should be fixed.